radix-engine/src/blueprints/pool/single_resource_pool/blueprint.rs

```diff
@@ -0,0 +1,408 @@
+use crate::blueprints::pool::single_resource_pool::*;
+use crate::errors::*;
+use crate::kernel::kernel_api::*;
+use native_sdk::modules::access_rules::*;
+use native_sdk::modules::metadata::*;
+use native_sdk::modules::royalty::*;
+use native_sdk::resource::*;
+use native_sdk::runtime::Runtime;
+use radix_engine_common::math::*;
+use radix_engine_common::prelude::*;
+use radix_engine_interface::api::node_modules::metadata::*;
+use radix_engine_interface::api::*;
+use radix_engine_interface::blueprints::pool::*;
+use radix_engine_interface::blueprints::resource::*;
+use radix_engine_interface::types::*;
+use radix_engine_interface::*;
+
+pub const SINGLE_RESOURCE_POOL_BLUEPRINT_IDENT: &'static str = "SingleResourcePool";
+
+pub struct SingleResourcePoolBlueprint;
+impl SingleResourcePoolBlueprint {
+    pub fn instantiate<Y>(
+        resource_address: ResourceAddress,
+        pool_manager_rule: AccessRule,
+        api: &mut Y,
+    ) -> Result<SingleResourcePoolInstantiateOutput, RuntimeError>
+    where
+        Y: ClientApi<RuntimeError> + KernelNodeApi,
+    {
+        // Validate that the resource is a fungible resource - a pool can't be created with non
+        // fungible resources.
+        let resource_manager = ResourceManager(resource_address);
+        if let ResourceType::NonFungible { .. } = resource_manager.resource_type(api)? {
+            Err(SingleResourcePoolError::NonFungibleResourcesAreNotAccepted { resource_address })?
+        }
+
+        // Allowing the component address of the pool - this will be used later for the component
+        // caller badge.
+        let node_id = api.kernel_allocate_node_id(EntityType::GlobalSingleResourcePool)?;
+        let address = GlobalAddress::new_or_panic(node_id.0);
+
+        let pool_unit_resource = {
+            let component_caller_badge = NonFungibleGlobalId::global_caller_badge(address);
+
+            let access_rules = btreemap!(
+                Mint => (
+                    rule!(require(component_caller_badge.clone())),
+                    AccessRule::DenyAll,
+                ),
+                Burn => (rule!(require(component_caller_badge)), AccessRule::DenyAll),
+                Recall => (AccessRule::DenyAll, AccessRule::DenyAll)
+            );
+
+            // TODO: Pool unit resource metadata - one things is needed to do this:
+            // 1- A fix for the issue with references so that we can have the component address of
+            //    the pool component in the metadata of the pool unit resource (currently results in
+            //    an error because we're passing a reference to a node that doesn't exist).
+
+            ResourceManager::new_fungible(18, Default::default(), access_rules, api)?.0
+        };
+
+        let object_id = {
+            let vault = Vault::create(resource_address, api)?;
```

Comment:
If we'll be storing the vault as a `Vault` and resource manager as a `ResourceManager` then it might be _slightly_ easier to just use `Vault::create` here since it returns a `Vault` whereas `resource_manager.new_empty_vault` returns an `Own`.